src/lib/variants.ts

```diff
@@ -0,0 +1,143 @@
+    excludedTagIds: any[];
+    excludedSections: string[];
+    alwaysAsk: boolean;
+    maxViews?: MaxViews;
+    userCohort: string;
+    isLiveBlog: boolean;
+    hasCountryName: boolean;
+    variants: Variant[];
+    highPriority: boolean;
+    useLocalViewLog: boolean;
+    articlesViewedSettings?: ArticlesViewedSettings;
+
+    // TODO check with Contributions Team as not in current examples so perhaps
+    // they are always 0 and 1?
+    audience: number;
+    audienceOffset?: number;
+}
+
+export interface EpicTests {
+    tests: Test[];
+}
+
+interface Filter {
+    id: string;
+    test: (test: Test, targeting: EpicTargeting) => boolean;
+}
+
+// https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/experiments/ab-core.js#L39
+const selectVariant = (test: Test, mvtId: number): Variant | undefined => {
+    // Calculate range of mvtIDs for variants and return first match
+    const lowest = mvtId * (test.audienceOffset || 0);
+    const highest = lowest + mvtId * test.audience;
+
+    if (mvtId > lowest && mvtId <= highest) {
+        return test.variants[mvtId % test.variants.length];
+    }
+
+    return undefined;
+};
+
+export const findVariant = (
+    data: EpicTests,
+    targeting: EpicTargeting,
+    mvtId: number,
+): Variant | undefined => {
+    // Also need to include canRun of individual variants (only relevant for
+    // manually configured tests).
+
+    // TODO include max view throttling here too
+    // https://github.com/guardian/frontend/blob/master/static/src/javascripts/projects/common/modules/commercial/contributions-utilities.js#L378
+    /* return (
+        (!initVariant.canRun || initVariant.canRun()) &&
+        meetsMaxViewsConditions &&
+        matchesCountryGroups &&
+        matchesTagsOrSections &&
+        noExcludedTags &&
+        notExcludedSection &&
+        copyIsValid()
+    ); */
+
+    const hasTags: Filter = {
+        id: 'hasTags',
+        test: (test, targeting) => {
+            if (test.tagIds.length < 1) {
+                return true;
+            }
+
+            const intersection = test.tagIds.filter(tagId =>
+                targeting.tags.map(tag => tag.id).includes(tagId),
+            );
+
+            return intersection.length > 0;
+        },
+    };
+
+    const hasSection: Filter = {
+        id: 'hasSection',
+        test: (test, targeting) => {
+            if (test.sections.length < 1) {
+                return true;
+            }
+
+            return test.sections.includes(targeting.sectionName);
+        },
+    };
+
+    const userInTest: Filter = {
+        id: 'userInTest',
+        test: (test, _) => selectVariant(test, mvtId) !== undefined,
+    };
+
+    const filters: Filter[] = [hasSection, hasTags, userInTest];
+    const test = data.tests.find(test =>
+        filters.every(filter => {
+            const got = filter.test(test, targeting);
+
+            if (!got && process.env.NODE_ENV !== 'production') {
```

Comment:
I'd like something to help with debugging, but I'm not sure this approach is the best way(!).

Reply:
@tjmw yeah a separate env var sounds good. Re comparing against Frontend, we could modify the Frontend existing test logic to also call out to our service and log any differences (while ignoring whatever we return)? That would give us a sense of any gaps using real production behaviour.

Reply:
I agree, being able to compare our testing logic against frontend would be really helpful, as it's fairly intricate stuff and possibly subject to the existence of good enough example variants in frontend.